Update dispatchEvent documentation for clarity (#43521)
mdn/content: files/en-us/web/api/eventtarget/dispatchevent/index.md

> [!NOTE]
> When calling this method, the {{domxref("Event.target")}} property is initialized to the current `EventTarget`.

> Unlike calling `dispatchEvent()` manually, "native" events are fired by the browser and dispatched asynchronously via the [event loop](/en-US/docs/Web/JavaScript/Reference/Execution_model). The dispatch process itself is similar in both cases and invokes event handlers _synchronously_. All applicable event handlers are called and return before `dispatchEvent()` returns.

Reviewer: Not only does each event handler get called in its stack frame (it doesn't make sense to not to); it's also called in its own job.

> And in that model: queued microtasks will run in-between, queued macrotasks will not run in-between, only after.

That's because other queued macrotasks necessarily have lower priority than these alraedy-queued event handlers. I'm not sure what that's proving.

I'm saying that there's no single job called "dispatch all event listeners". Basically, here's `dispatchEvent()`:

```js
function dispatchEvent(e) {
  for (const onclick of this.clickHandlers) {
    onclick(e);
  }
}
```

Here's the code that's run whenever a native click happens (this code itself can be executed synchronously or asynchronously—doesn't matter because there's no one to observe its synchronicity):

```js
for (const onclick of element.clickHandlers) {
  queueTask(() => onclick(e));
}
```

> Unlike calling `dispatchEvent()` manually, "native" events are fired by the browser and dispatched asynchronously via the [event loop](/en-US/docs/Web/JavaScript/Reference/Execution_model). The dispatch process itself is similar in both cases and invokes event handlers _synchronously_. All applicable event handlers are called and return before `dispatchEvent()` returns.

Author: I broadened my knowledge.
I also ran tests that show that rendering is "blocked" - deferred until all handlers run, even if making a visual change from one handler, and then blocking the thread from another. Even if making visual changes from within a queued microtask from within a handler.

As I now know about Blink as an outer host environment for Javascript execution threads - I'm pretty sure that the actual technical difference between native and manual is - manual triggers the dispatch "from within V8", native triggers it from "outside" - from Blink, schedules on the event loop, but it runs as a "single unit", not yielding back to the main event loop - or anyway it's "atomic". Microtasks run because there is also a microtasks checkpoint defined after each script execution (as well as at the end of an event loop iteration).

This is as I thought in the beginning, only with more actual implementation details.

I suggest this edit:

> Unlike calling `dispatchEvent()` manually, which triggers event handlers to run synchronously in the same stack frame - "native" events are dispatched by the browser, triggering each event handler to run on a separate stack execution frame; any queued microtasks will run after each event handler, but other asynchronous browser operations, such as rendering, will only run after all event handlers finished running.

> Unlike calling `dispatchEvent()` manually, "native" events are fired by the browser and dispatched asynchronously via the [event loop](/en-US/docs/Web/JavaScript/Reference/Execution_model). The dispatch process itself is similar in both cases and invokes event handlers _synchronously_. All applicable event handlers are called and return before `dispatchEvent()` returns.

Author: @dipikabh you're also okay with this?

> Unlike calling `dispatchEvent()` manually, "native" events are fired by the browser and dispatched asynchronously via the [event loop](/en-US/docs/Web/JavaScript/Reference/Execution_model). The dispatch process itself is similar in both cases and invokes event handlers _synchronously_. All applicable event handlers are called and return before `dispatchEvent()` returns.

Reviewer 2: Sorry @yuval-a, this fell off my radar.

Since @Josh-Cena has approved this latest version, I'm happy to go with it albeit with some tweaks and to break up the long sentence:

> Unlike calling `dispatchEvent()` manually, which triggers event handlers to run synchronously in the same stack frame, "native" events are dispatched by the browser, which triggers each event handler to run on a separate stack execution frame; any queued microtasks run after each event handler. However, other asynchronous browser operations, such as rendering, run only after all event handlers have finished running.

> Unlike calling `dispatchEvent()` manually, "native" events are fired by the browser and dispatched asynchronously via the [event loop](/en-US/docs/Web/JavaScript/Reference/Execution_model). The dispatch process itself is similar in both cases and invokes event handlers _synchronously_. All applicable event handlers are called and return before `dispatchEvent()` returns.

Reviewer 3: Not all native events are fired in their own task (and thus spawn a new stack per callback, which itself will enter the microtask checkpoint when cleaning). 
For instance the `<iframe>`'s load event will fire synchronously at document insertion if its `src` is  `about:blank`: https://jsfiddle.net/0cus135z/

> Unlike calling `dispatchEvent()` manually, "native" events are fired by the browser and dispatched asynchronously via the [event loop](/en-US/docs/Web/JavaScript/Reference/Execution_model). The dispatch process itself is similar in both cases and invokes event handlers _synchronously_. All applicable event handlers are called and return before `dispatchEvent()` returns.

Author: @Kaiido Interesting.
BUT... your code calls `document.body.append(i);` - which is **conceptually** like calling `element.click()` (it's a different kind of "dispatch" - it spawns from JS - like an "artificial click" - meaning it's not a "native event" "conformally" - it doesn't come from "outside" (e.g. an actual mouse click detected by the OS );

Note that once I put the `iframe` in the html (without `src`) - and even set `src` to `about:blank` from JS - the expected behavior happens.
See variant of your jsfiddle: https://jsfiddle.net/eqt29b6x/2/

Or it could be that the `createElement` of the iframe from JS is what's making the difference - this can make sense because in the first variation JS is the source of creation of the iframe, and in the second variation it's the HTML parser - which is in the host/Blink layer - so this also might be the reason for the `load` behaving differently between variations - it's like a "native load".
Either way I don't think it shows an exception.

Also I saw this:
https://github.com/whatwg/html/issues/4965
But not sure is related or even relevant anymore as my first paragraph above suggests.

I'm going to try to commit the PR with dipikabh phrasing from yesterday - in about two more days (to allow some more time for any other "possible exceptions").

> Unlike calling `dispatchEvent()` manually, "native" events are fired by the browser and dispatched asynchronously via the [event loop](/en-US/docs/Web/JavaScript/Reference/Execution_model). The dispatch process itself is similar in both cases and invokes event handlers _synchronously_. All applicable event handlers are called and return before `dispatchEvent()` returns.

Reviewer 3: I'm not sure everyone will understand this term that way. For me at least, a "native event" is an event fired from the UA itself, and not one generated by the user code that would be dispatched by a direct call to `dispatchEvent` (or by a wrapper like `HTMLElement.click()`).   

Other such events I know of, that would fall into *my* definition of "synchronous native events": `<form>`'s *formdata* and `<input>` *invalid* events when the form is submitted programmatically, `AbortSignal`'s *abort* event (even when tied by `AbortSignal.any()`), `<dialog>` before toggle when triggered from `close()`, *popstate*  when setting `location.hash`. But there may be many more in APIs I don't know. Events have to be fired from a task only when "the algorithm that triggers the event could be running on a different thread or process" (https://w3ctag.github.io/design-principles/)

Also, while I believe this might be a bug, in Firefox, *beforeinput*  and *input*  events do fire synchronously, even though they'd fall in *your* definition.

> Unlike calling `dispatchEvent()` manually, "native" events are fired by the browser and dispatched asynchronously via the [event loop](/en-US/docs/Web/JavaScript/Reference/Execution_model). The dispatch process itself is similar in both cases and invokes event handlers _synchronously_. All applicable event handlers are called and return before `dispatchEvent()` returns.

Author: I would define "native event" as "anything that does not originate from within the runtime of Javascript".
This also aligns with the quote you mention from the design specs - and also with your examples.

Meaning: "if you do it from JS - it's not a "native event"". I think this definition holds - also aligns with your examples - e.g. "submit with JS" vs. the browser's "native submit" (a button with type="submit" inside a form).

I would also argue that possibly any API implementation parts that ARE in JS actually call `dispatchEvent`.

"native events" are "system events" --- the obvious examples are things like mouse click - but are also true for something that happens in the "outer host browser level" - "browsery things".

The most important insight and anecdote I've learned from all of this is to remember that Javascript still keeps its original intention of acting as an "interface" to "outer" or "native" capabilities (it started as a "scripting language to interface with the browser" after all) - and those widened incredibly far over the years - but this truth still holds.

The mental model is seeing this single V8 thread as the "inside", and the host browser and beyond as the "outside". If it's triggered "programmatically" with JS === not "native".

> Unlike calling `dispatchEvent()` manually, "native" events are fired by the browser and dispatched asynchronously via the [event loop](/en-US/docs/Web/JavaScript/Reference/Execution_model). The dispatch process itself is similar in both cases and invokes event handlers _synchronously_. All applicable event handlers are called and return before `dispatchEvent()` returns.

Author: Come to think of it - there is no "official definition" in MDN of what a "native event" is (at least, I asked chat GPT to check and it says there isn't), while "fired by the user agent" IS sometimes used in specs, so maybe we should use:

> Unlike calling `dispatchEvent()` manually, which triggers event handlers to run synchronously in the same stack frame, "native" events - originating from the browser environment, triggers each event handler to run on a separate stack execution frame; any queued microtasks run after each event handler. However, other asynchronous browser operations, such as rendering, run only after all event handlers have finished running.

> Unlike calling `dispatchEvent()` manually, which triggers event handlers to run synchronously in the same stack frame, "native" events (dispatched by the browser) triggers each event handler to run on a separate stack execution frame; any queued microtasks run after each event handler. However, other asynchronous browser operations, such as rendering, run only after all event handlers have finished running.

Reviewer 4: @Josh-Cena @dipikabh Looking at this text ^^^ it seems pretty unhelpful to me - should I as a user of this method need to understand stack frames or microtask? No, no I should not - certainly not for this method where knowing this makes no difference.

This was also discussed in #43973, where we were converging towards something like this:

```suggestion
Unlike "native" events, which the browser fires by queuing a task on the [event loop](/en-US/docs/Web/JavaScript/Reference/Execution_model), `dispatchEvent()` invokes all applicable event handlers _synchronously_ before returning.
```

It is possibly not quite correct, but it captures the main point of the original, which is that these events are synchronous, whereas events generated by the browser are not.

Anyway, my take on this is that something simple here is better or deleting the text entirely, as largely irrelevant to readers.

I'll leave to you, and close the other issue as it was "late to the party".

Reviewer 4: ... or if you leave it to me, I'll just delete this section altogether :-)

Reviewer: The fundamental question is whether there's race—i.e., if it's possible for other jobs to execute in between two event handler executions.

Author: Following @hamishwillee comment about not being too "low-level-ish" (though I argue about his remark regarding microtasks - modern JS developers are familiar with the concept, especially as `queueMicrotask` is an official API, MDN also has a guide on it (added a link to it in my suggestion), plus the event loop is mentioned - it's just the same amount of "low-level", if not more) I ) and in the spirit of this paragraph being about "native events VS. user events", I now suggest changing to adding this note (while removing that paragraph)

```suggestion
> [!NOTE]
> For events dispatched with this method, the [`isTrusted` property of the Event object]
> (/en-US/docs/Web/API/Event/isTrusted) will be `false`. This is also true for events
> triggered programmatically by other means (for example, by calling `element.click()`).
> By contrast, events dispatched by the browser will have `isTrusted` set to `true`.
>
> Additionally, [microtasks](/en-US/docs/Web/API/HTML_DOM_API/Microtask_guide) scheduled
> from event handlers of an event dispatched programmatically will run only after all
> handlers have finished executing. For browser-dispatched events, microtasks may run
> between individual event listener invocations during the dispatch process.
>
> Other browser work, such as rendering, does not occur until after the entire event dispatch and handling has completed, in both cases.
```

Author: > @Josh-Cena @dipikabh Looking at this text ^^^ it seems pretty unhelpful to me - should I as a user of this method need to understand stack frames or microtask? No, no I should not - certainly not for this method where knowing this makes no difference.
> 
> This was also discussed in #43973, where we were converging towards something like this:
> 
> It is possibly not quite correct, but it captures the main point of the original, which is that these events are synchronous, whereas events generated by the browser are not.
> 
> Anyway, my take on this is that something simple here is better or deleting the text entirely, as largely irrelevant to readers.
> 
> I'll leave to you, and close the other issue as it was "late to the party".

@hamishwillee I disagree about the comment about should they be familiar with what a `microtask` is. Yes - they should, `queueMicrotask` is official API. MDN has this guide: https://developer.mozilla.org/en-US/docs/Web/API/HTML_DOM_API/Microtask_guide.
The event loop is mentioned - which is an even "lower-level" concept.

I think the questions regarding about whether this paragraph should exist, and if so - is this the right place, are:
* Would a reader of the `dispatchEvent` page will likely be interested in a quick summary of the differences between "custom and native"? I say: yes.
* What would represent the practical differences? I say: mainly the microtasks difference, the non render between handlers, then I also think the `isTrusted`.

On the other hand - your last suggestion is short and concise - and does "fix" the technical semantic "mistake" that triggered me to open this PR.

Reviewer 4: > The fundamental question is whether there's race—i.e., if it's possible for other jobs to execute in between two event handler executions.

@Josh-Cena If this is the fundamental question/problem then IMO should be addressed directly rather than highlighting that some things are sync and others are async. I.e. start with the problem.

@yuval-a As a general principle I think we will have to agree to disagree on microtasks - most programmers most of the time do not need to care about that level of detail.  Of course it may be that this is a case where the problem can't be explained without them. Fortunately I'm not the reviewer, so I don't need to assess that.

The  paragraph on isTrusted seems useful YOu might consider a few words on why it matters (in addition to the link)

Author: @hamishwillee I thought about just mentioning Promises but it sounds too partial.
Anyway, I think the subject of "do we go there?" about Microtasks - is like a "standards/specs" question. I'm not a long-running contributor to MDN - I came here because I was triggered by the phrasing of that paragraph - but do MDN have a formal "standards" document? like a "style guide"? If not, it should - I was going per what I know already exists. But anyway I emphasis again that I think it's very legit to use that term;
1. As I mentioned there's actually an official global method with that term in its name (for me that alone is enough to justify it).
2. Microtasks is the de facto mechanism of "asynchronicity" in Javascript - before it existed it was just "offload to the host browser" for async. To the best of my knowledge adding it was a design choice spawned from: "We need a way to allow these "clean-up/just-before" tasks inside Javascript. It's used for Promise resolve callbacks, but also for Custom Elements reactions, MutationObserver callbacks and likely other stuff. It's a thing in Javascript, like, a major thing. 

Anyway, a compromise would be to mention Promises instead and add "Microtasks" parenthesised. Also, I'm leaning more and more towards a note that explains the two different types of events.

This is my next suggestion:

```suggestion
> [!NOTE]
> Events dispatched programmatically using `dispatchEvent()` (as well as by other
> means, such as calling [`click()`](/en-US/docs/Web/API/HTMLElement/click))
> are informally known as "synthetic events", while events dispatched by
> the browser are often referred to as "native events".
>
> Synthetic events (including custom events) have their
> [`isTrusted`](/en-US/docs/Web/API/Event/isTrusted) property set to `false`, while
> browser-generated events have it set to `true`. This allows applications 
> and browser features to distinguish between browser-generated events 
> and script-dispatched events.
>
> Additionally, [Promise](en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise) 
> callbacks (and other microtasks) scheduled from event handlers of
> synthetic events will run only after all event handlers have completed running. For
> browser-dispatched events, microtasks may run between individual event handler 
> invocations during dispatch. Other browser work, such as rendering, occurs only after
> the entire event dispatch has completed in both cases.
```

Reviewer 4: @yuval-a Thanks. I think this is getting htere ^^^. I wouldn't necessarily have (all of) it as a note because it isn't really all aside material. You would want to tidy this bit "scheduled from event handlers of "synthetic events will run only after all event handlers have completed running." because it compares event handlers and event handlers without differentiation of the types you mean.

But I am leaving this to @dipikabh and @Josh-Cena

Reviewer 2: My 2 cents here: The line just before Line 22 is already a note callout. Also, the proposed text in https://github.com/mdn/content/pull/43521#discussion_r3205015741 would introduce a large block of text before "Syntax" - @hamishwillee is that normal for the API reference area, or should more details like this go in a "Description" section instead?

@yuval-a I am okay going with following the earlier suggestion from Hamish, it focuses on the original issue and keeps it minimal:

"Unlike "native" events, which the browser fires by queuing a task on the [event loop](/en-US/docs/Web/JavaScript/Reference/Execution_model), `dispatchEvent()` invokes all applicable event handlers _synchronously_ before returning."

@Josh-Cena, can you confirm if this version looks okay to you?

Reviewer 4: FWIW perfect is the enemy of good. The minimal fix might be imperfect, but we seem to be stalled here.

Author: @dipikabh I am okay with the phrasing you presented in your comment as it represents loyalfully the technicality. Sorry for the delay in the response.

Author: I commited a change with your phrasing verbatim, plus a sentence about `isTrusted`. I don't think it needs further explanation, because it's mostly just that - to flag a differentiation between 'native' and 'synthetic' (plus there's a link to the page about it).

> Unlike "native" events, which the browser fires by queuing a task on the [event loop](/en-US/docs/Web/API/Document_Object_Model/Events#creating_and_dispatching_events), `dispatchEvent()` invokes all applicable event handlers synchronously before returning. [Event](/en-US/docs/Web/API/Event) objects for native events have the read-only [`isTrusted`](/en-US/docs/Web/API/Event/isTrusted) property set to `true`, while events dispatched with `dispatchEvent()` have it set to `false`.

Reviewer 2: - Any reason for changing the link? The originally linked page looks more appropriate for "event loop". In fact, we could link to this section `/en-US/docs/Web/JavaScript/Reference/Execution_model#job_queue_and_event_loop`
- Tweaked the 2nd sentence

```suggestion
Unlike "native" events, which the browser fires by queuing a task on the [event loop](/en-US/docs/Web/JavaScript/Reference/Execution_model#job_queue_and_event_loop), `dispatchEvent()` invokes all applicable event handlers synchronously before returning. The read-only [`isTrusted`](/en-US/docs/Web/API/Event/isTrusted) property is `true` for native events and `false` for events dispatched with `dispatchEvent()`.
```
